examples/4_secret_storage.py

```diff
@@ -0,0 +1,377 @@
+        self, secret: bytes, team: str, namespace: str, entryKey: str
+    ) -> keybase1.KVGetResult:
+        h = await self.hmac_key(secret, entryKey)
+        res = await self.kvstore.get(team, namespace, h)
+        res.entry_key = h
+        return res
+
+    async def list_entrykeys(
+        self, secret: bytes, team: str, namespace: str
+    ) -> keybase1.KVListEntryResult:
+        res = await self.kvstore.list_entrykeys(team, namespace)
+        if res.entry_keys:
+            for e in res.entry_keys:
+                if not e.entry_key.startswith("_"):
+                    get_res = await self.kvstore.get(team, namespace, e.entry_key)
+                    e.entry_key = json.loads(get_res.entry_value)[
+                        self.KEY_KEY
+                    ]  # modify
+        return res
+
+
+class RentalHandler:
+    """
+    RentalHandler handles commands sent via chat to use the team key-value store.
+
+    RentalHandler listens to chat messages of the form:
+    `!rental {reserve|unreserve} <tool> <YYYY-MM-DD>`
+    `!rental {lookup|add|remove} <tool>`
+    `!rental list` // lists all tools
+
+    It expects properly formatted commands.
+
+    RentalHandler uses a cache to keep track of the most recently fetched value and
+    revision for each key. To handle concurrent updates, it attempts to update
+    with the most recently fetched revision + 1; if it fails, it does a "get"
+    and updates the cache, and asks the user to retry if they still want to
+    do their update.
+
+    RentalHandler also maintains a cache of the per-team per-namespace hmac secrets
+    it has previously fetched (stored with the special entryKey "_secret"),
+    which it uses to HMAC entryKeys. RentalHandler uses the SecretKeyKVStoreClient
+    to store entryKeys' HMACs instead of the plaintext entryKeys.
+   """
+
+    MSG_PREFIX = "!rental"
+    NAMESPACE = "rental"
+    SECRET_NUM_BYTES = 32
+    SECRET_KEY = "_secret"
+
+    def __init__(self):
+        # self.cache = {tool: {"revision": int, "info": {} or None}}
+        self.cache: Dict[Any, Any] = {}
+        self.secrets: Dict[str, Dict[str, bytes]] = {}  # {team: {namespace: secret}}
+
+    async def load_secret(self, bot, team, namespace):
+        if team not in self.secrets or namespace not in self.secrets[team]:
+            secret = secrets.token_bytes(self.SECRET_NUM_BYTES)
+            try:
+                # we don't expect self.SECRET_KEY's revision > 0
+                res: keybase1.KVPutResult = await bot.kvstore.put(
+                    team, namespace, self.SECRET_KEY, bytes_to_str(secret), revision=1
+                )
+            except RevisionError:
+                res = await bot.kvstore.get(team, namespace, self.SECRET_KEY)
+                secret = str_to_bytes(res.entry_value)
+            if team not in self.secrets:
+                self.secrets[team] = {}
+            self.secrets[team][namespace] = secret
+        return self.secrets[team][namespace]
```

Comment:
Another thing we should make explicit here is that the hmac secrets are never updated: if I get removed from a team (and hack into keybase servers), I can still check if a specific entry name has been created since I was removed from the team (although I would not be able to decrypt the entry values). 
It is subtle but worth making explicit because people might not think of this.
cc @xgess on this.

Reply:
ugh. i don't think you're wrong on this point. ok, maybe we can add a comment at the bottom of the file that explains in a little detail 1) how the secret client obfuscates key names, 2) drawbacks (the server can still see access patterns, and previous members of the team who explicitly save the secrets could collude with the keybase server to see entry keys but not updated entry values). i'd like this comment to be pretty separated from the basic "what am i looking at" comment though.

Reply:
>I can still check if a specific entry name has been created since I was removed from the team
this is if you somehow got direct access to the db, right? keybase server checks if you have access to a team before executing a kvstore cli command.

Reply:
correct. you would need database access because the keybase server will not allow access to someone who is not a current member of the team. this is one of those things that's a guarantee we provide but we can't actually prove it to the user. it's tough to convey that in a readable comment, but yeah.

Reply:
trying to convey this in next commit